File: python_test_crate/tests/05_enter_exit.py
from typing import Any, Optional, List, Dict, Callable

class EnterExitFrameEvent:
    def __init__(self, message: str, parameters):
        self._message = message
        self._parameters = parameters
        self._return = None


class EnterExitCompartment:
    def __init__(self, state: str, parent_compartment = None):
        self.state = state
        self.state_args = {}
        self.state_vars = {}
        self.enter_args = {}
        self.exit_args = {}
        self.forward_event = None
        self.parent_compartment = parent_compartment

    def copy(self) -> 'EnterExitCompartment':
        c = EnterExitCompartment(self.state, self.parent_compartment)
        c.state_args = self.state_args.copy()
        c.state_vars = self.state_vars.copy()
        c.enter_args = self.enter_args.copy()
        c.exit_args = self.exit_args.copy()
        c.forward_event = self.forward_event
        return c


class EnterExit:
    def __init__(self):
        self._state_stack = []
        self._return_value = None
        self.log =         []
        self.__compartment = EnterExitCompartment("Off")
        self.__next_compartment = None
        __frame_event = EnterExitFrameEvent("$>", None)
        self.__kernel(__frame_event)

    def __kernel(self, __e):
        # Route event to current state
        self.__router(__e)
        # Process any pending transition
        while self.__next_compartment is not None:
            next_compartment = self.__next_compartment
            self.__next_compartment = None
            # Exit current state
            exit_event = EnterExitFrameEvent("<$", self.__compartment.exit_args)
            self.__router(exit_event)
            # Switch to new compartment
            self.__compartment = next_compartment
            # Enter new state (or forward event)
            if next_compartment.forward_event is None:
                enter_event = EnterExitFrameEvent("$>", self.__compartment.enter_args)
                self.__router(enter_event)
            else:
                # Forward event to new state
                forward_event = next_compartment.forward_event
                next_compartment.forward_event = None
                if forward_event._message == "$>":
                    # Forwarding enter event - just send it
                    self.__router(forward_event)
                else:
                    # Forwarding other event - send $> first, then forward
                    enter_event = EnterExitFrameEvent("$>", self.__compartment.enter_args)
                    self.__router(enter_event)
                    self.__router(forward_event)

    def __router(self, __e):
        state_name = self.__compartment.state
        handler_name = f"_state_{state_name}"
        handler = getattr(self, handler_name, None)
        if handler:
            handler(__e)

    def __transition(self, next_compartment):
        self.__next_compartment = next_compartment

    def toggle(self):
        __e = EnterExitFrameEvent("toggle", None)
        self.__kernel(__e)

    def get_log(self) -> list:
        self._return_value = None
        __e = EnterExitFrameEvent("get_log", None)
        self.__kernel(__e)
        return self._return_value

    def _state_Off(self, __e):
        if __e._message == "<$":
            self.log.append("exit:Off")
            print("Exiting Off state")
        elif __e._message == "$>":
            self.log.append("enter:Off")
            print("Entered Off state")
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return
        elif __e._message == "toggle":
            __compartment = EnterExitCompartment("On", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)

    def _state_On(self, __e):
        if __e._message == "<$":
            self.log.append("exit:On")
            print("Exiting On state")
        elif __e._message == "$>":
            self.log.append("enter:On")
            print("Entered On state")
        elif __e._message == "get_log":
            self._return_value = self.log
            __e._return = self._return_value
            return
        elif __e._message == "toggle":
            __compartment = EnterExitCompartment("Off", parent_compartment=self.__compartment.copy())
            self.__transition(__compartment)


def main():
    print("=== Test 05: Enter/Exit Handlers ===")
    s = EnterExit()

    # Initial enter should have been called
    log = s.get_log()
    assert "enter:Off" in log, f"Expected 'enter:Off' in log, got {log}"
    print(f"After construction: {log}")

    # Toggle to On - should exit Off, enter On
    s.toggle()
    log = s.get_log()
    assert "exit:Off" in log, f"Expected 'exit:Off' in log, got {log}"
    assert "enter:On" in log, f"Expected 'enter:On' in log, got {log}"
    print(f"After toggle to On: {log}")

    # Toggle back to Off - should exit On, enter Off
    s.toggle()
    log = s.get_log()
    assert log.count("enter:Off") == 2, f"Expected 2 'enter:Off' entries, got {log}"
    assert "exit:On" in log, f"Expected 'exit:On' in log, got {log}"
    print(f"After toggle to Off: {log}")

    print("PASS: Enter/Exit handlers work correctly")

if __name__ == '__main__':
    main()

File: rust_test_crate/tests/05_enter_exit.rs
use std::collections::HashMap;

#[derive(Clone, Debug)]
struct EnterExitFrameEvent {
    message: String,
}

impl EnterExitFrameEvent {
    fn new(message: &str) -> Self {
        Self { message: message.to_string() }
    }
}

#[derive(Clone)]
struct EnterExitCompartment {
    state: String,
    state_vars: std::collections::HashMap<String, i32>,
    forward_event: Option<EnterExitFrameEvent>,
}

impl EnterExitCompartment {
    fn new(state: &str) -> Self {
        Self {
            state: state.to_string(),
            state_vars: std::collections::HashMap::new(),
            forward_event: None,
        }
    }
}

#[derive(Clone)]
enum EnterExitStateContext {
    Off,
    On,
    Empty,
}

impl Default for EnterExitStateContext {
    fn default() -> Self {
        EnterExitStateContext::Off
    }
}

pub struct EnterExit {
    _state_stack: Vec<(String, EnterExitStateContext)>,
    __compartment: EnterExitCompartment,
    __next_compartment: Option<EnterExitCompartment>,
    log: Vec<String>,
}

impl EnterExit {
    pub fn new() -> Self {
        let mut this = Self {
            _state_stack: vec![],
            log: Vec::new(),
            __compartment: EnterExitCompartment::new("Off"),
            __next_compartment: None,
        };
let __frame_event = EnterExitFrameEvent::new("$>");
this.__kernel(__frame_event);
        this
    }

    fn __kernel(&mut self, __e: EnterExitFrameEvent) {
// Route event to current state
self.__router(&__e);
// Process any pending transition
while self.__next_compartment.is_some() {
    let next_compartment = self.__next_compartment.take().unwrap();
    // Exit current state
    let exit_event = EnterExitFrameEvent::new("$<");
    self.__router(&exit_event);
    // Switch to new compartment
    self.__compartment = next_compartment;
    // Enter new state (or forward event)
    if self.__compartment.forward_event.is_none() {
        let enter_event = EnterExitFrameEvent::new("$>");
        self.__router(&enter_event);
    } else {
        // Forward event to new state
        let forward_event = self.__compartment.forward_event.take().unwrap();
        if forward_event.message == "$>" {
            // Forwarding enter event - just send it
            self.__router(&forward_event);
        } else {
            // Forwarding other event - send $> first, then forward
            let enter_event = EnterExitFrameEvent::new("$>");
            self.__router(&enter_event);
            self.__router(&forward_event);
        }
    }
}
    }

    fn __router(&mut self, __e: &EnterExitFrameEvent) {
match self.__compartment.state.as_str() {
    "Off" => self._state_Off(__e),
    "On" => self._state_On(__e),
    _ => {}
}
    }

    fn __transition(&mut self, next_compartment: EnterExitCompartment) {
self.__next_compartment = Some(next_compartment);
    }

    fn _state_stack_push(&mut self) {
let state_context = match self.__compartment.state.as_str() {
    "Off" => EnterExitStateContext::Off,
    "On" => EnterExitStateContext::On,
    _ => EnterExitStateContext::Empty,
};
self._state_stack.push((self.__compartment.state.clone(), state_context));
    }

    fn _state_stack_pop(&mut self) {
let (saved_state, state_context) = self._state_stack.pop().unwrap();
let exit_event = EnterExitFrameEvent::new("$<");
self.__router(&exit_event);
self.__compartment.state = saved_state;
match state_context {
    EnterExitStateContext::Off => {}
    EnterExitStateContext::On => {}
    EnterExitStateContext::Empty => {}
}
    }

    pub fn toggle(&mut self) {
let __e = EnterExitFrameEvent::new("toggle");
self.__kernel(__e);
    }

    pub fn get_log(&mut self) -> Vec<String> {
let __e = EnterExitFrameEvent::new("get_log");
match self.__compartment.state.as_str() {
            "Off" => self._s_Off_get_log(&__e),
            "On" => self._s_On_get_log(&__e),
            _ => Default::default(),
        }
    }

    fn _state_Off(&mut self, __e: &EnterExitFrameEvent) {
match __e.message.as_str() {
    "$<" => { self._s_Off_exit(__e); }
    "$>" => { self._s_Off_enter(__e); }
    "get_log" => { self._s_Off_get_log(__e); }
    "toggle" => { self._s_Off_toggle(__e); }
    _ => {}
}
    }

    fn _state_On(&mut self, __e: &EnterExitFrameEvent) {
match __e.message.as_str() {
    "$<" => { self._s_On_exit(__e); }
    "$>" => { self._s_On_enter(__e); }
    "get_log" => { self._s_On_get_log(__e); }
    "toggle" => { self._s_On_toggle(__e); }
    _ => {}
}
    }

    fn _s_Off_exit(&mut self, __e: &EnterExitFrameEvent) {
self.log.push("exit:Off".to_string());
println!("Exiting Off state");
    }

    fn _s_Off_get_log(&mut self, __e: &EnterExitFrameEvent) -> Vec<String> {
self.log.clone()
    }

    fn _s_Off_enter(&mut self, __e: &EnterExitFrameEvent) {
self.log.push("enter:Off".to_string());
println!("Entered Off state");
    }

    fn _s_Off_toggle(&mut self, __e: &EnterExitFrameEvent) {
self.__transition(EnterExitCompartment::new("On"));
    }

    fn _s_On_toggle(&mut self, __e: &EnterExitFrameEvent) {
self.__transition(EnterExitCompartment::new("Off"));
    }

    fn _s_On_get_log(&mut self, __e: &EnterExitFrameEvent) -> Vec<String> {
self.log.clone()
    }

    fn _s_On_enter(&mut self, __e: &EnterExitFrameEvent) {
self.log.push("enter:On".to_string());
println!("Entered On state");
    }

    fn _s_On_exit(&mut self, __e: &EnterExitFrameEvent) {
self.log.push("exit:On".to_string());
println!("Exiting On state");
    }
}


fn main() {
    println!("=== Test 05: Enter/Exit Handlers ===");
    let mut s = EnterExit::new();

    // Initial enter should have been called
    let log = s.get_log();
    assert!(log.contains(&"enter:Off".to_string()), "Expected 'enter:Off' in log, got {:?}", log);
    println!("After construction: {:?}", log);

    // Toggle to On - should exit Off, enter On
    s.toggle();
    let log = s.get_log();
    assert!(log.contains(&"exit:Off".to_string()), "Expected 'exit:Off' in log, got {:?}", log);
    assert!(log.contains(&"enter:On".to_string()), "Expected 'enter:On' in log, got {:?}", log);
    println!("After toggle to On: {:?}", log);

    // Toggle back to Off - should exit On, enter Off
    s.toggle();
    let log = s.get_log();
    let enter_off_count = log.iter().filter(|s| *s == "enter:Off").count();
    assert_eq!(enter_off_count, 2, "Expected 2 'enter:Off' entries, got {:?}", log);
    assert!(log.contains(&"exit:On".to_string()), "Expected 'exit:On' in log, got {:?}", log);
    println!("After toggle to Off: {:?}", log);

    println!("PASS: Enter/Exit handlers work correctly");
}
